designs/2023-relative-cache/README.md

```diff
@@ -0,0 +1,203 @@
+directory. Typically, eslint is always run from the same place, so using a relative path should not cause a problem.
+
+Note that the obscure hash value is still there, and is now Ak0sovs. This hash value will need to be calculated using
+relative paths only. 
+
+``` 
+[
+  {
+    "./src/formatter.ts": "1",
+    "./src/vite-env.d.ts": "2"
+  },
+  {
+    "size": 2584,
+    "mtime": 1695823503788,
+    "results": "3",
+    "hashOfConfig": "4"
+  },
+  {
+    "size": 75,
+    "mtime": 1695814197580,
+    "results": "5",
+    "hashOfConfig": "4"
+  },
+  {
+    "filePath": "6",
+    "messages": "7",
+    "suppressedMessages": "8",
+    "errorCount": 0,
+    "fatalErrorCount": 0,
+    "warningCount": 0,
+    "fixableErrorCount": 0,
+    "fixableWarningCount": 0
+  },
+  "Ak0sovs",
+  {
+    "filePath": "9",
+    "messages": "10",
+    "suppressedMessages": "11",
+    "errorCount": 0,
+    "fatalErrorCount": 0,
+    "warningCount": 0,
+    "fixableErrorCount": 0,
+    "fixableWarningCount": 0
+  },
+  "./src/formatter.ts",
+  [],
+  [],
+  "./src/vite-env.d.ts",
+  [],
+  []
+]
+```
+
+### Adding the command line parameter
+- `conf/default-cli-options.js`: add the property `shareableCache` with a default of false. It should be put in the section with the other cache variables, below `cacheStrategy`. 
+- `docs/src/use/command-line-interface.md`: Add an explanation of the `shareable-cache` variable with the other cache variables: "By default, an eslintcache contains full file paths and thus cannot readily be shared across developers or ci machines. "False" is that default. "True" changes the internal storage to store a relative path to the process execution directory, thus making the eslintcache shareable across developers and ci machines. . If you change this setting, you must regenerate your entire eslint cache."
+- `eslint-helpers.js`: add `shareableCache` variable (set to false) to the `processOptions`, and make sure it must be a boolean. 
+- `lib/options.js`: Add an option `shareable-cache` of type Boolean with a nice description for people to read: By default, an eslintcache contains full file paths and thus cannot readily be shared across developers or ci machines. "False" is that default. "True" changes the internal storage to store a relative path to the process execution directory, thus making the eslintcache shareable across developers and ci machines. . If you change this setting, you must regenerate your entire eslint cache.
+- `lib/cli.js`: Add the `shareableCache` option, defaulted to false, in the `translateOptions `function. 
+
+### Changing cache file serialization
+- `lib/cli-engine/lint-result-cache.js`: Add the properties `cwd` and `shareableCache` to the `LintResultCache` class. `cwd` is a string, the current working directory. `shareableCache` is a boolean, the result of the user's passed in command line parameter. Use these values in two main places: 
```

Comment:
I think we should clarify what we mean by the current working directory that will be passed to the `LintResultCache` class. Is it `process.cwd()`, or `cwd` that was passed to ESLint/FlatESLint constructor? In the latter case, is there a way to pass `cwd` to `file-entry-cache` or we could end up calculating paths differently?

Reply:
Another question is does `file-entry-cache` support relative paths? All tests seem to be with absolute paths:

https://github.com/jaredwray/file-entry-cache/blob/master/test/specs/cache.js

Reply:
It will. I am just going to push up the code review sometime today. I have a complete and working branch on this task, and most of  your questions wuold be resolved by the code review/better answered looking at the code changes.

Reply:
Don't forget to answer the question regarding what `cwd` is from the previous comment.

Reply:
> Is it `process.cwd()`, or `cwd` that was passed to ESLint/FlatESLint constructor?

When created in cli-engine.js, it's options.cwd. It is also referred to as 'cwd' in the code. 
When created in flat-eslint.js, it's processedOptions.cwd.

Reply:
Ok. I have updated the code, and the flat-eslint.js file no longer exists in my code branch. I see a lot of the code duplication around parameter parsing has been removed. I am updating my RFC accordingly.

Reply:
I think the question whether `file-entry-cache` supports relative paths and what happens when `cwd` passed to ESLint is different than `process.cwd()` hasn't been addressed.

Reply:
To clarify the problem with cwd: When we store lint results for a file, `file-entry-cache` reads that file to get the data (metadata or content, depending on the cache strategy) it will use next time we use cache to determine if the file has been modified in the meantime and thus if its cache entry should be invalidated. So, when we pass a file path, `file-entry-cache` needs to find that file on the disk. If the path is relative, the question is how `file-entry-cache` resolves it. If we pass a path that is relative to `options.cwd`, and `file-entry-cache` resolves it relative to `process.cwd()` (likely), then it won't find the file when `options.cwd` is different than `process.cwd()`.

Another important question is whether `file-entry-cache` officially supports relative paths, or maybe that works just by chance but could break without notice.

Reply:
@jaredwray 👋 Would you be able to help us answer this question: does `file-entry-cache` expect or require absolute paths? Is there any reason you could think of that would prevent it from using relative paths?

Reply:
@nzakas - I dont believe it would matter if you use relative or absolute. Want me to create some tests to validate?

Reply:
@jaredwray that would be awesome, thank you. 🙏

Reply:
great. I will add in some unit tests over the next couple days to the main project (https://github.com/jaredwray/file-entry-cache) to make sure it doesnt matter. 

From what I am gathering we want to be able to do:

```js
const path = require('path');
const fileEntryCache = require('file-entry-cache');
const relativePath = './relative-cache';
const absolutePath = path.join(process.cwd(), relativePath);
const cacheRelative = fileEntryCache.create('cacheName', [relativePath]);
const cacheAbsolute = fileEntryCache.create('cacheName', [absolutePath]);

//then we run tests to validate each!

```

Is this what we want to verify or something more?

Reply:
We'd like to verify that we can both write to and read from the cache with both absolute and relative paths.

The fear from the thread above, is that we may be relying on unintentional behavior that could change unexpectedly. So by adding these tests, we can verify both that this works and that it won't accidentally stop working in the future.

Reply:
@mdjermanovic do you have an example that would satisfy your question?

Reply:
@nzakas  - I am writing the tests that will check that you can read and write with relative and absolute to validate this over the next 72 hours and will post the results here.

Reply:
Awesome, thank you!

Reply:
@jaredwray just checking back to see if you had a chance to look into those tests?

Reply:
I wrote some tests using `cwd()` and it seems to be working just great. Are you seeing any issues? If so, can you give me an easy example of how to reproduce so we can fix it?

Reply:
Here is a run down from our tests:
* create (works with both)
* createFromFile (works with both)

The issue is that the keys are absolute paths so what happens is that when you do something like:
* .getUpdatedFiles(files);
* .hasFileChanged(files[0]);

etc... These are all based on the absolute path for the key. Does that make sense?

Reply:
@nzakas Ok. So in this case you would want to pass in something like `../folder/file_name.txt` instead of the full path. Correct?

Reply:
Yep. It requires an absolute path. I could work to do a release where it doesn't. Is that the best option for you all?

```javascript
  it('getFileDescriptor with relative paths', function () {
    cache = fileEntryCache.create('testCache');
    var file = '../fixtures/' + fixtureFiles[0].name;
    cache = fileEntryCache.createFromFile('../fixtures/.eslintcache');
    expect(cache.getFileDescriptor(file).notFound).to.be.false;
    cache.reconcile();
    deleteFileSync(file);
    expect(cache.getFileDescriptor(file).notFound).to.be.true;
  });
```

This failed on the first expect as the file should exist.

Reply:
@nzakas - I pushed a new pull request that detects if it is absolute or relative and then does the conversion on `removeEntry` and `getFileDescriptor` which are the two main culprits. Can you all test if this fixes it? If it does, then I am happy to do a release. 

https://github.com/jaredwray/file-entry-cache/pull/46

Reply:
For our purposes, what we're looking for is the ability to reference files relative to the cache file itself rather than from the cwd.

For example, if we have a `.eslintcache` file, we'd like to be able to say `getFileDescriptor("foo/bar.js")` and have that understood to be relative to the `.eslintcache` location. 

The goal is to have the cache file continue to work even if we move it to a different system where the directory structure above it might be different but the directory structure below it is the same.

Reply:
makes sense. I am wondering if we add in a rootPath parameter that you can set or change if you move the file and it keeps the consistency? Something along the lines of:

```javascript
cache = fileEntryCache.create('testCache');
var cache = fileEntryCache.createFromFile('/fixtures/.eslintcache');
var info1 = cache.getFileDescriptor('foo/bar.js');
cache.reconsile();
cache.rootPath = '/fixtures/.eslintcache1';
cache.reconsile();  // this updates the base path so that you can 
var info2 = cache.getFileDescriptor('foo/bar.js');
expect(info1).to.equal(info2);
```

Reply:
I'm curious if we need a root path at all? As in, do you need to store anything other than the relative path as it was passed?

Reply:
I was assuming that the path would change to a new one so we would need to know where the new one would be.

Reply:
do you want to setup some time to do a video call? I can do it this weekend if needed as I don't want to be the bottleneck on this solution. :-)

Reply:
Here's the list of methods we're using:

* API `create()`
* `FileEntryCache#getFileDescriptor()`
* `FileEntryCache#reconcile()`

Reply:
@jaredwray We'd like to make cache files shareable between different systems, where the project is possibly cloned/copied into a different directory.

For example, currently, if someone has the following directory structure and makes a cache file:

```
/
└── projects/
    └── my-app/
        ├── .eslintcache
        ├── eslint.config.js
        └── my-file.js
```

the cache entry for `my-file.js` is saved with `"/projects/my-app/my-file.js"` key.

Then, if `my-app/` is copied to another system into directory `work/`:

```
/
└── work/
    └── my-app/
        ├── .eslintcache
        ├── eslint.config.js
        └── my-file.js
```

the cache file is unusable because there's no `/projects/my-app/my-file.js` file on that system.

We've had multiple requests to address this by using relative paths instead of absolute paths as keys in the cache file:

* https://github.com/eslint/eslint/issues/16493
* https://github.com/eslint/eslint/issues/16741
* https://github.com/eslint/eslint/issues/14571
* https://github.com/eslint/eslint/issues/17514

Reply:
A detail I think we're still unsure about is which location the paths should be relative to - what would make the most sense for ESLint users, and then what could `file-entry-cache` support?

Possible candidates:

1. `process.cwd()`
2. `cwd` option passed to ESLint (the same as `process.cwd()` when ESLint is used via CLI)
3. Location of the cache file
4. Location of the `eslint.config.js` file (`basePath`)

Option 1 (and consequently option 2 when ESLint is used via CLI) could be problematic because ESLint can be run from different locations so file paths in the cache might end up being inconsistent.

@nzakas per https://github.com/eslint/rfcs/pull/114#discussion_r1514962901, seems that you think that option 3 makes the most sense?

Reply:
> ```js
> cache = fileEntryCache.create('testCache');
> var cache = fileEntryCache.createFromFile('/fixtures/.eslintcache');
> var info1 = cache.getFileDescriptor('foo/bar.js');
> cache.reconsile();
> cache.rootPath = '/fixtures/.eslintcache1';
> cache.reconsile();  // this updates the base path so that you can 
> var info2 = cache.getFileDescriptor('foo/bar.js');
> expect(info1).to.equal(info2);
> ```

We don't intend to support moving the cache file during the execution. What we'd like to support is something like this:

Step 1: Run eslint in directory `/foo` and store data for file `bar.js` (`/foo/bar.js`) in the cache

```js
const fileEntryCache = require("file-entry-cache");

const cache = fileEntryCache.create("/foo/.eslintcache", undefined, /* usecheckSum = */ true);

const fileDescriptor = cache.getFileDescriptor("bar.js");

// store some data
fileDescriptor.meta.results = { something: 42 };

cache.reconcile();
```

Step 2: Copy the entire content (including the cache file) of `/foo` into `/baz`

Step 3: Run eslint in directory `/baz` and use the stored data for file `bar.js` (now `/baz/bar.js`) from the cache.

```js
const fileEntryCache = require("file-entry-cache");

const cache = fileEntryCache.create("/baz/.eslintcache", undefined, /* usecheckSum = */ true);

const fileDescriptor = cache.getFileDescriptor("bar.js");

// use the data
if (!fileDescriptor.changed) { // fileDescriptor.changed should be `false` because the content of `bar.js` hasn't changed
    console.log(fileDescriptor.meta.results); // should be { something: 42 }
}
```

Reply:
This makes sense and I plan to have a pull request on Monday (4/1) that can handle this scenario so you can validate with it first. Is that ok?

Reply:
> Ok. I have updated the code, and the flat-eslint.js file no longer exists in my code branch. I see a lot of the code duplication around parameter parsing has been removed. I am updating my RFC accordingly.

In ESLint 9, eslint.js has been renamed to legacy-eslint.js and flat-eslint.js is now named eslint.js.

Reply:
@jaredwray are you still interested in submitting a pull request? Just let us know if you need any help.

Reply:
@fasttime - I made the changes to support the relative path and wanted to confirm that is the change you need to make it work with file-entry-cache.

Reply:
@fasttime - https://github.com/jaredwray/file-entry-cache/pull/46

Reply:
> @fasttime - [jaredwray/file-entry-cache#46](https://github.com/jaredwray/file-entry-cache/pull/46)

Thanks! Now it would be great if someone could draft a PR in ESLint to test these changes.

Reply:
> jaredwray/file-entry-cache#46

@jaredwray I tried this with passing relative paths from ESLint, and it didn't work for the scenario described in https://github.com/eslint/rfcs/pull/114#discussion_r1532521752. It looks like the change is that, when a relative path is passed to `getFileDescriptor()`, it converts it to an absolute path, but at the end cache file still stores absolute paths?

Reply:
@nzakas, @fasttime there's still a design question for us: which location the paths should be relative to? (https://github.com/eslint/rfcs/pull/114#discussion_r1532605561)

Reply:
> @nzakas, @fasttime there's still a design question for us: which location the paths should be relative to? ([#114 (comment)](https://github.com/eslint/rfcs/pull/114#discussion_r1532605561))

That should be probably the directory containing the cache file, i.e. option 3 in the comment linked above. Option 4 would also work but with the additional requirement that a config file must exist, which is not needed to work with the ESLint API or to use the cache.

Reply:
I think ideally it would be relative to the location of the cache file, however, if someone is using `--cache-location`, then I think the expectation may be that the paths are relative to cwd?

Reply:
@mdjermanovic @fasttime what do you think of [my last comment](https://github.com/eslint/rfcs/pull/114#discussion_r1616179161)?

Reply:
> I think ideally it would be relative to the location of the cache file, however, if someone is using `--cache-location`, then I think the expectation may be that the paths are relative to cwd?

When `--cache-location` is not specified, the location of the cache file is `cwd` by default. So if there is an expectation that the paths are relative to `cwd` when `--cache-location` is used, the same expectation should be probably there when `--cache-location` is not used. Do you think that `cwd` would be a better base location to store relative paths?

Reply:
> Do you think that `cwd` would be a better base location to store relative paths?

Do you mean in all cases? If so, the answer is no. I think it makes the most sense to have it relative to the cache file when `--cache-location` is not used.

Reply:
> > Do you think that `cwd` would be a better base location to store relative paths?
> 
> Do you mean in all cases? If so, the answer is no. I think it makes the most sense to have it relative to the cache file when `--cache-location` is not used.



> > Do you think that `cwd` would be a better base location to store relative paths?
> 
> Do you mean in all cases? If so, the answer is no. I think it makes the most sense to have it relative to the cache file when `--cache-location` is not used.



> > Do you think that `cwd` would be a better base location to store relative paths?
> 
> Do you mean in all cases? If so, the answer is no. I think it makes the most sense to have it relative to the cache file when `--cache-location` is not used.



> > Do you think that `cwd` would be a better base location to store relative paths?
> 
> Do you mean in all cases? If so, the answer is no. I think it makes the most sense to have it relative to the cache file when `--cache-location` is not used.

I'm asking because when `--cache-location` is not set, the cache file location should be already `cwd`. I think the advantage of keeping the paths relative to the cache file directory is that they don't change if the current directory changes, for example in a monorepo. Do you have a particular scenario in mind where storing paths relative to `cwd` would be advantageous?

Reply:
Yeah, as @fasttime noticed, when `--cache-location` is not used, the cache file defaults to `.eslintcache` _in cwd_. So if the logic would be to use the location of the cache file when `--cache-location` is not used and cwd when `--cache-location`  is used, then it's basically the same as - always use cwd.

Reply:
I think we should just make a decision and see what happens when people test with it. This PR has been open since October of last year, and it seems like this is the last thing to decide on, so I don't think there's value in drawing this out further when we have a binary decision to make.

So I'd like to propose that we say the paths are always relative to location of the `.eslintcache` file.

Reply:
> So I'd like to propose that we say the paths are always relative to location of the `.eslintcache` file.

I agree.